File: ice0208/trie/208.py
# https://www.youtube.com/watch?v=oobqoCJlHA0
# 처음 만들어보는 자료구조라 영상을 참고하면서 만들었다.
# 나중에 복습필요.


class TrieNode:
    def __init__(self):
        self.children = {}
        self.endOfWord = False


class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        cur = self.root

        for c in word:
            if c not in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]
        cur.endOfWord = True

    def search(self, word: str) -> bool:
        cur = self.root

        for c in word:
            if c not in cur.children:
                return False
            cur = cur.children[c]
        return cur.endOfWord

    def startsWith(self, prefix: str) -> bool:
        cur = self.root

        for c in prefix:
            if c not in cur.children:
                return False
            cur = cur.children[c]
        return True

File: moonpoison/trie/208.cpp
#include<bits/stdc++.h>
using namespace std;

class TrieNode {
public:
    TrieNode *child[26];
    bool isWord;
    TrieNode() 
    {
        isWord = false;
        for (auto &a : child) a = nullptr;
    }
};
class Trie {
    TrieNode* root;
public:
    Trie() 
    {
        root = new TrieNode();
    }
    void insert(string s) 
    {
        TrieNode *p = root;
        for (auto &a : s) 
        {
            int i = a - 'a';
            if (!p->child[i]) p->child[i] = new TrieNode();
            p = p->child[i];
        }
        p->isWord = true;
    }
    bool search(string key, bool prefix=false) 
    {
        TrieNode *p = root;
        for (auto &a : key) 
        {
            int i = a - 'a';
            if (!p->child[i]) return false;
            p = p->child[i];
        }
        if (prefix==false) return p->isWord;
        return true;
    }
    bool startsWith(string prefix) 
    {
        return search(prefix, true);
    }
};